File: Builder/UnitTests.VB/UnitTestInitializer.vb

'Code taken from http://blogs.iis.net/yigalatz/archive/2011/03/31/unit-tests-should-not-debug-assert.aspx
Imports System.Diagnostics

<TestClass>
Public Class UnitTestInitializer

    Private Shared _initialized As Boolean = False
    Private Shared _lock As New Object()

    <AssemblyInitialize>
    Public Shared Sub Initialize(context As TestContext)
        SyncLock _lock
            If _initialized Then
                Return
            End If

            Dim removeListener As TraceListener = Nothing
            For Each listener As TraceListener In Debug.Listeners
                If TypeOf listener Is DefaultTraceListener Then
                    removeListener = listener
                    Exit For
                End If
            Next
            Debug.Listeners.Remove(removeListener)
        End SyncLock
        Debug.Listeners.Add(FailOnAssert.GetInstance())
    End Sub

End Class

File: Builder/UnitTests/UnitTestInitializer.cs

using System.Diagnostics;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Questify.Builder.UnitTests
{
    /// <summary>
    /// This class is here for a construct to ensure that when a test Debug.Asserts the test fails.
    /// </summary>
    [TestClass]
    public class UnitTestInitializer
    {
        private static bool _initialized = false;
        private static object _lock = new object();

        [AssemblyInitialize]
        public static void Initialize(TestContext context)
        {
            lock (_lock)
            {
                if (_initialized)
                {
                    return;
                }

                TraceListener removeListener = null;
                foreach (TraceListener listener in Debug.Listeners)
                {
                    if (listener is DefaultTraceListener)
                    {
                        removeListener = listener;
                        break;
                    }
                }
                Debug.Listeners.Remove(removeListener);
                Debug.Listeners.Add(FailOnAssert.GetInstance());
            }
        }
    }

}